keps/sig-cloud-provider/1959-service-lb-class-annotation/README.md

```diff
@@ -0,0 +1,203 @@
+support similar functionality for Services of Type=LoadBalancer. Introducing a new resource like
+`ServiceClass` is probably not worthwhile given that there are new APIs already in development.
+This KEP proposes a light-weight approach for Service Type=LoadBalancer by introducing a Service
+annotation `service.kubernetes.io/load-balancer-class`.
+
+## Motivation
+
+The main use-case for this feature is being able to support multiple Service Type=LoadBalancer
+implementations in a cluster, as different workloads may want to leverage different loadbalancer
+providers based on efficiency, availability, cost and other factors.
+
+For example, a cluster admin may want to use a public load balancer from a cloud provider
+for workloads that must be assigned a publically routable address, but they may want to
+enable a lower-cost solution for workloads that are only internally accessible.
+
+### Goals
+
+* allow users to opt-out of the Service Type=LoadBalancer implementation by the cloud provider.
+* allow multiple implementations of Service Type=LoadBalancer in a given cluster.
+
+### Non-Goals
+
+* performance improvements for Service Type=LoadBalancer.
+* changing any other existing behaviors for Service Type=LoadBalancer aside from being able
+to disabling it from the cloud provider.
+
+## Proposal
+
+This KEP proposes to add a new Service annotation `service.kubernetes.io/load-balancer-class`
+that allows for multiple implementations of Service Type=LoadBalancer in a cluster.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin:
+* I want to use my cloud provider's public load balancer service for applications that require
+public ingress.
+* I want to use my own load balancing solution for any applications that only talk internally
+within my own network because I want to save costs.
+
+#### Story 2
+
+As an application developer:
+* I MUST use a hardware-based loadbalancer for certain applications due to specific protocols
+only available there.
+* I want to use the cloud provider's default load balancer for any applications that do not
+rely on protocols from hardware load balancers.
+
+### Risks and Mitigations
+
+Many Service Type=LoadBalancer implementations today support a lot of knobs via annotations already.
+Introducing yet another annotation for Service Type=LoadBalancer is not ideal, but this is better than
+every cloud provider supporting their own "skip this Service" annotation.
+
+## Design Details
+
+Introduce a new Service annotation `service.kubernetes.io/load-balancer-name`.
+
+If the loadbalancer class annotation is not set, the existing cloud provider
+will assume ownership of the Service Type=LoadBalancer resource. This is required
+to not break existing clusters that assume Service Type=LoadBalancer is always
+managed by the cloud provider.
+
+A new method `SupportedLoadBalancerClassNames()` will be added to the cloud provider
```

Comment:
This is maybe overkill? We can maybe skip this and assume any Service without the annotation matches against the cloud provider, no need to check for class names within the cloud provider implementations, even if multiple cloud providers today support different types of LBs (e.g. AWS ELB vs NLB).